# Conditions



Attaching the needed libraries:


```r
library(rlang, warn.conflicts = FALSE)
library(testthat, warn.conflicts = FALSE)
```

### Exercises 8.2.4

---

**Q1.** Write a wrapper around `file.remove()` that throws an error if the file to be deleted does not exist.

**A1.** Let's first create a wrapper function around `file.remove()` that throws an error if the file to be deleted does not exist.


```r
fileRemove <- function(...) {
  existing_files <- fs::file_exists(...)

  if (!all(existing_files)) {
    stop(
      cat(
        "The following files to be deleted don't exist:",
        names(existing_files[!existing_files]),
        sep = "\n"
      )
    )
  }

  file.remove(...)
}
```

Let's first create a file that we can delete immediately.


```r
fs::file_create("random.R")
```

The function should fail if there are any other files provided that don't exist:


```r
fileRemove(c("random.R", "XYZ.csv"))
#> The following files to be deleted don't exist:
#> XYZ.csv
#> Error in fileRemove(c("random.R", "XYZ.csv")):
```

But does work as expected when the file exists:


```r
fileRemove("random.R")
#> [1] TRUE
```

---

**Q2.** What does the `appendLF` argument to `message()` do? How is it related to `cat()`?

**A2.** As mentioned in the docs for `message()`, `appendLF` argument decides:

> should messages given as a character string have a newline appended?

- If `TRUE`, a final newline is regarded as part of the message:


```r
message("Beetle", appendLF = TRUE)
#> Beetle
message("Juice", appendLF = TRUE)
#> Juice
```

- If `FALSE`, messages will be concatenated:


```r
message("Beetle", appendLF = FALSE)
#> Beetle
message("Juice", appendLF = FALSE)
#> Juice
```

On the other hand, `cat()` converts its arguments to character vectors and concatenates them to a single character vector by default:


```r
cat("Beetle")
#> Beetle
cat("Juice")
#> Juice
```

In order to get output `message()`-like default behavior, we can set `sep = "\n"`:


```r
cat("Beetle", sep = "\n")
#> Beetle
cat("Juice", sep = "\n")
#> Juice
```

---

### Exercises 8.4.5

---

**Q1.** What extra information does the condition generated by `abort()` contain compared to the condition generated by `stop()` i.e. what's the difference between these two objects? Read the help for `?abort` to learn more.


```r
catch_cnd(stop("An error"))
catch_cnd(abort("An error"))
```

**A1.** Contrasting these two lists, we can see that `rlang::abort()` contains two additional pieces of information:

- `trace`: A traceback capturing the sequence of calls that lead to the current function 
- `parent`: Information about another condition used as a parent to create a chained condition.


```r
library(rlang)

stopInfo <- catch_cnd(stop("An error"))
abortInfo <- catch_cnd(abort("An error"))

str(stopInfo)
#> List of 2
#>  $ message: chr "An error"
#>  $ call   : language force(expr)
#>  - attr(*, "class")= chr [1:3] "simpleError" "error" "condition"

str(abortInfo)
#> List of 4
#>  $ message: chr "An error"
#>  $ trace  :Classes 'rlang_trace', 'rlib_trace', 'tbl' and 'data.frame':	8 obs. of  6 variables:
#>   ..$ call       :List of 8
#>   .. ..$ : language catch_cnd(abort("An error"))
#>   .. ..$ : language eval_bare(rlang::expr(tryCatch(!!!handlers, {     force(expr) ...
#>   .. ..$ : language tryCatch(condition = `<fn>`, {     force(expr) ...
#>   .. ..$ : language tryCatchList(expr, classes, parentenv, handlers)
#>   .. ..$ : language tryCatchOne(expr, names, parentenv, handlers[[1L]])
#>   .. ..$ : language doTryCatch(return(expr), name, parentenv, handler)
#>   .. ..$ : language force(expr)
#>   .. ..$ : language abort("An error")
#>   ..$ parent     : int [1:8] 0 1 1 3 4 5 1 0
#>   ..$ visible    : logi [1:8] FALSE FALSE FALSE FALSE FALSE FALSE ...
#>   ..$ namespace  : chr [1:8] "rlang" "rlang" "base" "base" ...
#>   ..$ scope      : chr [1:8] "::" "::" "::" "local" ...
#>   ..$ error_frame: logi [1:8] FALSE FALSE FALSE FALSE FALSE FALSE ...
#>   ..- attr(*, "version")= int 2
#>  $ parent : NULL
#>  $ call   : NULL
#>  - attr(*, "class")= chr [1:3] "rlang_error" "error" "condition"
```

---

**Q2.** Predict the results of evaluating the following code


```r
show_condition <- function(code) {
  tryCatch(
    error = function(cnd) "error",
    warning = function(cnd) "warning",
    message = function(cnd) "message",
    {
      code
      NULL
    }
  )
}

show_condition(stop("!"))
show_condition(10)
show_condition(warning("?!"))
show_condition({
  10
  message("?")
  warning("?!")
})
```

**A2.** Correctly predicted ðŸ˜‰


```r
show_condition <- function(code) {
  tryCatch(
    error = function(cnd) "error",
    warning = function(cnd) "warning",
    message = function(cnd) "message",
    {
      code
      NULL
    }
  )
}

show_condition(stop("!"))
#> [1] "error"
show_condition(10)
#> NULL
show_condition(warning("?!"))
#> [1] "warning"
show_condition({
  10
  message("?")
  warning("?!")
})
#> [1] "message"
```

The last piece of code is the challenging one and it illustrates how `tryCatch()` works. From its docs:

> When several handlers are supplied in a single tryCatch then the first one is considered more recent than the second.

---

**Q3.** Explain the results of running this code:


```r
withCallingHandlers(
  message = function(cnd) message("b"),
  withCallingHandlers(
    message = function(cnd) message("a"),
    message("c")
  )
)
#> b
#> a
#> b
#> c
```

**A3.** The surprising part of this output is the `b` before the last `c`.

This happens because the inner calling handler doesn't handle the message, so it bubbles up to the outer calling handler.

---

**Q4.** Read the source code for `catch_cnd()` and explain how it works.

**A4.** Let's look at the source code for `catch_cnd()`:


```r
rlang::catch_cnd
#> function (expr, classes = "condition") 
#> {
#>     stopifnot(is_character(classes))
#>     handlers <- rep_named(classes, list(identity))
#>     eval_bare(rlang::expr(tryCatch(!!!handlers, {
#>         force(expr)
#>         return(NULL)
#>     })))
#> }
#> <bytecode: 0x0000000033cdfbc8>
#> <environment: namespace:rlang>
```

As mentioned in the function docs:

> This is a small wrapper around `tryCatch()` that captures any condition signalled while evaluating its argument.

The `classes` argument allows a character vector of condition classes to catch, and the complex tidy evaluation generates the necessary condition (if there is any; otherwise `NULL`).


```r
catch_cnd(10)
#> NULL

catch_cnd(abort(message = "an error", class = "my class"))
#> <error/my class>
#> Error:
#> ! an error
#> ---
#> Backtrace:

catch_cnd(abort("an error"), classes = "message")
#> Error:
#> ! an error
```

---

**Q5.** How could you rewrite `show_condition()` to use a single handler?

**A5.** The source code for `rlang::catch_cond()` gives us a clue as to how we can do this.

Conditions also have a `class` attribute, and we can use it to determine which handler will match the condition.


```r
show_condition2 <- function(code) {
  tryCatch(
    condition = function(cnd) {
      if (inherits(cnd, "error")) {
        return("error")
      }
      if (inherits(cnd, "warning")) {
        return("warning")
      }
      if (inherits(cnd, "message")) {
        return("message")
      }
    },
    {
      code
      NULL
    }
  )
}
```

Let's try this new version with the examples from original version:


```r
show_condition2(stop("!"))
#> [1] "error"
show_condition2(10)
#> NULL
show_condition2(warning("?!"))
#> [1] "warning"
show_condition2({
  10
  message("?")
  warning("?!")
})
#> [1] "message"
```

---

### Exercises 8.5.4

---

**Q1.** Inside a package, it's occasionally useful to check that a package is installed before using it. Write a function that checks if a package is installed (with `requireNamespace("pkg", quietly = FALSE))` and if not, throws a custom condition that includes the package name in the metadata.

**A1.** Here is the desired function:


```r
abort_missing_package <- function(pkg) {
  msg <- glue::glue("Problem loading `{pkg}` package, which is missing and must be installed.")

  abort("error_missing_package",
    message = msg,
    pkg = pkg
  )
}

check_if_pkg_installed <- function(pkg) {
  tryCatch(
    requireNamespace(pkg, quietly = FALSE),
    error_missing_package = abort_missing_package(pkg)
  )
}

check_if_pkg_installed("xyz123")
#> Error in `abort_missing_package()`:
#> ! Problem loading `xyz123` package, which is missing and must be installed.
check_if_pkg_installed("dplyr")
#> Error in `abort_missing_package()`:
#> ! Problem loading `dplyr` package, which is missing and must be installed.
```

For a reference, also see the source code for following functions:

- `rlang::is_installed()`
- `insight::check_if_installed()`

---

**Q2.** Inside a package you often need to stop with an error when something is not right. Other packages that depend on your package might be tempted to check these errors in their unit tests. How could you help these packages to avoid relying on the error message which is part of the user interface rather than the API and might change without notice?

**A2.** As an example, let's say that another package developer wanted to use the `check_if_pkg_installed()` function that we just wrote.

So the developer using it in their own package can write a unit test like this:


```r
expect_error(
  check_if_pkg_installed("xyz123"),
  "Problem loading `xyz123` package, which is missing and must be installed."
)
```

To dissuade developers from having to rely on error messages to check for errors, we can instead provide a custom condition, which can be used for unit testing instead:


```r
e <- catch_cnd(check_if_pkg_installed("xyz123"))

inherits(e, "error_missing_package")
#> [1] TRUE
```

So that the unit test could be:


```r
expect_s3_class(e, "error_missing_package")
```

This test wouldn't fail even if we decided to change the exact message.

---

### Exercises 8.6.6

---

**Q1.** Create `suppressConditions()` that works like `suppressMessages()` and `suppressWarnings()` but suppresses everything. Think carefully about how you should handle errors.

**A1.** The following function suppresses/muffles both warnings and messages, except error. If errors are present, it doesn't produce an error, but returns a list instead that contains the error message.


```r
suppressConditions <- function(expr) {
  tryCatch(
    error = function(cnd) {
      list(result = NULL, error = cnd)
    },
    withCallingHandlers(
      message = function(cnd) {
        if (inherits(cnd, "message")) tryInvokeRestart("muffleMessage")
      },
      warning = function(cnd) {
        if (inherits(cnd, "warning")) tryInvokeRestart("muffleWarning")
      },
      expr
    )
  )
}
```

Let's try out and see if this works as expected:


```r
suppressConditions(1)
#> [1] 1

suppressConditions({
  warning("I'm warning you")
  message("I'm messaging you")
})

suppressConditions({
  stop("I'm stopping this")
})
#> $result
#> NULL
#> 
#> $error
#> <simpleError in withCallingHandlers(message = function(cnd) {    if (inherits(cnd, "message"))         tryInvokeRestart("muffleMessage")}, warning = function(cnd) {    if (inherits(cnd, "warning"))         tryInvokeRestart("muffleWarning")}, expr): I'm stopping this>
```

---

**Q2.** Compare the following two implementations of `message2error()`. What is the main advantage of `withCallingHandlers()` in this scenario? (Hint: look carefully at the traceback.)


```r
message2error <- function(code) {
  withCallingHandlers(code, message = function(e) stop(e))
}
message2error <- function(code) {
  tryCatch(code, message = function(e) stop(e))
}
```

**A2.** With `withCallingHandlers()`, the condition handler is called from the signaling function itself, and, therefore, provides a more detailed call stack.



```r
message2error1 <- function(code) {
  withCallingHandlers(code, message = function(e) stop("error"))
}

message2error1({
  1
  message("hidden error")
  NULL
})
#> Error in (function (e) : error

traceback()
#> 9: stop("error") at #2
#> 8: (function (e)
#>    stop("error"))(list(message = "hidden error\n",
#>      call = message("hidden error")))
#> 7: signalCondition(cond)
#> 6: doWithOneRestart(return(expr), restart)
#> 5: withOneRestart(expr, restarts[[1L]])
#> 4: withRestarts({
#>        signalCondition(cond)
#>        defaultHandler(cond)
#>    }, muffleMessage = function() NULL)
#> 3: message("hidden error") at #1
#> 2: withCallingHandlers(code,
#>      message = function(e) stop("error")) at #2
#> 1: message2error1({
#>        1
#>        message("hidden error")
#>        NULL
#>    })
```

With `tryCatch()`, the signalling function terminates when a condition is raised, and so it doesn't provide as detailed call stack.


```r
message2error2 <- function(code) {
  tryCatch(code, message = function(e) (stop("error")))
}

message2error2({
  1
  stop("hidden error")
  NULL
})
#> Error in value[[3L]](cond) : error

traceback()
#> 6: stop("error") at #2
#> 5: value[[3L]](cond)
#> 4: tryCatchOne(expr, names, parentenv, handlers[[1L]])
#> 3: tryCatchList(expr, classes, parentenv, handlers)
#> 2: tryCatch(code, message = function(e) (stop("error"))) at #2
#> 1: message2error2({
#>        1
#>        message("hidden error")
#>        NULL
#>    })
```

---

**Q3.** How would you modify the `catch_cnds()` definition if you wanted to recreate the original intermingling of warnings and messages?

**A3.** Actually, you won't have to modify anything about the function defined in the chapter, since it supports this out of the box. 

So nothing additional to do here^[The best kind of exercise there is!]! ðŸ˜…


```r
catch_cnds <- function(expr) {
  conds <- list()
  add_cond <- function(cnd) {
    conds <<- append(conds, list(cnd))
    cnd_muffle(cnd)
  }

  withCallingHandlers(
    message = add_cond,
    warning = add_cond,
    expr
  )

  conds
}

catch_cnds({
  inform("a")
  warn("b")
  inform("c")
})
#> [[1]]
#> <message/rlang_message>
#> Message:
#> a
#> 
#> [[2]]
#> <warning/rlang_warning>
#> Warning:
#> b
#> 
#> [[3]]
#> <message/rlang_message>
#> Message:
#> c
```

---

**Q4.**  Why is catching interrupts dangerous? Run this code to find out.


```r
bottles_of_beer <- function(i = 99) {
  message(
    "There are ", i, " bottles of beer on the wall, ",
    i, " bottles of beer."
  )
  while (i > 0) {
    tryCatch(
      Sys.sleep(1),
      interrupt = function(err) {
        i <<- i - 1
        if (i > 0) {
          message(
            "Take one down, pass it around, ", i,
            " bottle", if (i > 1) "s", " of beer on the wall."
          )
        }
      }
    )
  }
  message(
    "No more bottles of beer on the wall, ",
    "no more bottles of beer."
  )
}
```

**A4.** Because this function catches the `interrupt` and works with it, there is no way to stop `bottles_of_beer()`, because the way you would usually stop it by using `interrupt`!


```r
bottles_of_beer()
#> There are 99 bottles of beer on the wall, 99 bottles of beer.
#> Take one down, pass it around, 98 bottles of beer on the wall.
#> Take one down, pass it around, 97 bottles of beer on the wall.
#> Take one down, pass it around, 96 bottles of beer on the wall.
#> Take one down, pass it around, 95 bottles of beer on the wall.
#> Take one down, pass it around, 94 bottles of beer on the wall.
#> Take one down, pass it around, 93 bottles of beer on the wall.
#> Take one down, pass it around, 92 bottles of beer on the wall.
#> Take one down, pass it around, 91 bottles of beer on the wall.
#> ...
```

In RStudio IDE, you can snap out of this loop by terminating the R session.

But, this shows why catching `interrupts` is dangerous and can result in poor user experience.

---

## Session information


```r
sessioninfo::session_info(include_base = TRUE)
#> - Session info -------------------------------------------
#>  setting  value
#>  version  R version 4.1.3 (2022-03-10)
#>  os       Windows 10 x64 (build 22000)
#>  system   x86_64, mingw32
#>  ui       RTerm
#>  language (EN)
#>  collate  English_United Kingdom.1252
#>  ctype    English_United Kingdom.1252
#>  tz       Europe/Berlin
#>  date     2022-08-20
#>  pandoc   2.19 @ C:/PROGRA~1/Pandoc/ (via rmarkdown)
#> 
#> - Packages -----------------------------------------------
#>  ! package     * version    date (UTC) lib source
#>    base        * 4.1.3      2022-03-10 [?] local
#>    bookdown      0.28       2022-08-09 [1] CRAN (R 4.1.3)
#>    brio          1.1.3      2021-11-30 [1] CRAN (R 4.1.2)
#>    bslib         0.4.0      2022-07-16 [1] CRAN (R 4.1.3)
#>    cachem        1.0.6      2021-08-19 [1] CRAN (R 4.1.1)
#>    cli           3.3.0      2022-04-25 [1] CRAN (R 4.1.3)
#>  P compiler      4.1.3      2022-03-10 [2] local
#>    crayon        1.5.1      2022-03-26 [1] CRAN (R 4.1.3)
#>  P datasets    * 4.1.3      2022-03-10 [2] local
#>    desc          1.4.1      2022-03-06 [1] CRAN (R 4.1.2)
#>    digest        0.6.29     2021-12-01 [1] CRAN (R 4.1.2)
#>    downlit       0.4.2      2022-07-05 [1] CRAN (R 4.1.3)
#>    evaluate      0.16       2022-08-09 [1] CRAN (R 4.1.3)
#>    fansi         1.0.3      2022-03-24 [1] CRAN (R 4.1.3)
#>    fastmap       1.1.0      2021-01-25 [1] CRAN (R 4.1.1)
#>    fs            1.5.2      2021-12-08 [1] CRAN (R 4.1.2)
#>    glue          1.6.2      2022-02-24 [1] CRAN (R 4.1.2)
#>  P graphics    * 4.1.3      2022-03-10 [2] local
#>  P grDevices   * 4.1.3      2022-03-10 [2] local
#>    htmltools     0.5.3      2022-07-18 [1] CRAN (R 4.1.3)
#>    jquerylib     0.1.4      2021-04-26 [1] CRAN (R 4.1.1)
#>    jsonlite      1.8.0      2022-02-22 [1] CRAN (R 4.1.2)
#>    knitr         1.39.9     2022-08-18 [1] Github (yihui/knitr@9e36e9c)
#>    lifecycle     1.0.1      2021-09-24 [1] CRAN (R 4.1.1)
#>    magrittr    * 2.0.3      2022-03-30 [1] CRAN (R 4.1.3)
#>    memoise       2.0.1      2021-11-26 [1] CRAN (R 4.1.2)
#>  P methods     * 4.1.3      2022-03-10 [2] local
#>    pillar        1.8.1      2022-08-19 [1] CRAN (R 4.1.3)
#>    pkgload       1.3.0      2022-06-27 [1] CRAN (R 4.1.3)
#>    R6            2.5.1.9000 2022-08-04 [1] Github (r-lib/R6@87d5e45)
#>    rlang       * 1.0.4      2022-07-12 [1] CRAN (R 4.1.3)
#>    rmarkdown     2.15.1     2022-08-18 [1] Github (rstudio/rmarkdown@b86f18b)
#>    rprojroot     2.0.3      2022-04-02 [1] CRAN (R 4.1.3)
#>    rstudioapi    0.13       2020-11-12 [1] CRAN (R 4.1.1)
#>    sass          0.4.2      2022-07-16 [1] CRAN (R 4.1.3)
#>    sessioninfo   1.2.2      2021-12-06 [1] CRAN (R 4.1.2)
#>  P stats       * 4.1.3      2022-03-10 [2] local
#>    stringi       1.7.8      2022-07-11 [1] CRAN (R 4.1.3)
#>    stringr       1.4.0      2019-02-10 [1] CRAN (R 4.1.2)
#>    testthat    * 3.1.4      2022-04-26 [1] CRAN (R 4.1.3)
#>  P tools         4.1.3      2022-03-10 [2] local
#>    utf8          1.2.2      2021-07-24 [1] CRAN (R 4.1.1)
#>  P utils       * 4.1.3      2022-03-10 [2] local
#>    vctrs         0.4.1      2022-04-13 [1] CRAN (R 4.1.3)
#>    winch         0.0.8      2022-03-17 [1] CRAN (R 4.1.3)
#>    withr         2.5.0      2022-03-03 [1] CRAN (R 4.1.2)
#>    xfun          0.32       2022-08-10 [1] CRAN (R 4.1.3)
#>    xml2          1.3.3      2021-11-30 [1] CRAN (R 4.1.2)
#>    yaml          2.3.5      2022-02-21 [1] CRAN (R 4.1.2)
#> 
#>  [1] C:/Users/IndrajeetPatil/Documents/R/win-library/4.1
#>  [2] C:/Program Files/R/R-4.1.3/library
#> 
#>  P -- Loaded and on-disk path mismatch.
#> 
#> ----------------------------------------------------------
```
